geopandas/tools/overlay.py

```diff
@@ -176,3 +176,143 @@ def overlay(df1, df2, how, use_sindex=True):
 
     # Return geodataframe with new indices
     return GeoDataFrame(collection, index=range(len(collection)))
+
+def overlay(df1, df2, how='intersection', reproject=True, use_sindex=None, **kwargs):
+    """Perform spatial overlay between two polygons.
+
+    Currently only supports data GeoDataFrames with polygons.
+    Implements several methods that are all effectively subsets of
+    the union.
+
+    Parameters
+    ----------
+    df1 : GeoDataFrame with MultiPolygon or Polygon geometry column
+    df2 : GeoDataFrame with MultiPolygon or Polygon geometry column
+    how : string
+        Method of spatial overlay: 'intersection', 'union',
+        'identity', 'symmetric_difference' or 'difference'.
+    reproject : boolean, default True
+        If GeoDataFrames do not have same projection, reproject
+        df2 to same projection of df1 before performing overlay
+
+    Returns
+    -------
+    df : GeoDataFrame
+        GeoDataFrame with new set of polygons and attributes
+        resulting from the overlay
+
+    """
+    if use_sindex is not None:
+        print('use_sindex is deprecated. If you are trying to use the old "overlay" function use "overlay_slow".')
+
+    # Allowed operations
+    allowed_hows = [
+        'intersection',
+        'union',
+        'identity',
+        'symmetric_difference',
+        'difference',  # aka erase
+    ]
+    # Error Messages
+    if how not in allowed_hows:
+        raise ValueError("`how` was \"%s\" but is expected to be in %s" % \
+            (how, allowed_hows))
+
+    if isinstance(df1, GeoSeries) or isinstance(df2, GeoSeries):
+        raise NotImplementedError("overlay currently only implemented for GeoDataFrames")
+
+    if (df1.geom_type.apply(lambda x: x in ['Polygon', 'MultiPolygon']).sum()!=len(df1.index) or 
+        df2.geom_type.apply(lambda x: x in ['Polygon', 'MultiPolygon']).sum()!=len(df2.index)):
+        raise TypeError("overlay only takes GeoDataFrames with (multi)polygon geometries") 
+
+    # Computations
+    df1 = df1.copy()
+    df2 = df2.copy()
+    df1['geometry'] = df1.geometry.buffer(0)
+    df2['geometry'] = df2.geometry.buffer(0)
```

Comment:
The buffering here is to fix possible 'invalid' geometries? 
Should this be optional, or should we warn the user that this is done? (I don't know to what extent this can have an influence on your geometries in case it was not needed)

Reply:
Yes, that is the idea. Not sure what the best strategy is. My experience is that it has no effect on geometries that are valid and as far as I can tell it resolves any issues behind the ``invalid`` geometries. So I guess we can go either way...a warning would be ok or just make it optional. The issues I would consider in making the decision are: 

1. Not sure if this step has speed implications or not. I know checking for whether the geometries are valid takes time. Doing this buffer usually is fast (although I imagine it will depend on the size of the dataframe, I have not profiled this).

2. When this fails due to invalid geometries, users do not get a message that will let them know the invalidity of geometries is behind the problem (as far as I remember the error message is a bit obscure and I am not sure if we can provide a better message without testing whether it is invalid geometries that cause it to fail). So that may cause users to blame ``gpd`` and report problems, when the problem is not really a ``gpd`` issue, but a data issue. Doing it by default (with a warning) may overcome this. Users can always choose to not have it done.

Reply:
One additional thing we may think about setting as default behavior when creating the copy of ``df1`` and ``df2`` is resetting the index. In this case I would favor a warning. Not sure why exactly but if one subsets a dataframe and then creates an overlay operation things either do not work and an error is produced or the output is wrong. I have not been able to figure out why this happens.